libdnf5/rpm/package_query.cpp

```diff
@@ -159,8 +160,14 @@ Id what_upgrades(libdnf5::solv::RpmPool & spool, const Solvable * solvable) {
         if (updated->repo != spool->installed || updated->name != solvable->name) {
             continue;
         }
-        if (updated->arch != solvable->arch && updated->arch != ARCH_NOARCH && solvable->arch != ARCH_NOARCH) {
-            continue;
+        if (updated->arch != solvable->arch) {
+            if (solvable->arch == ARCH_SRC || solvable->arch == ARCH_NOSRC || updated->arch == ARCH_SRC ||
+                updated->arch == ARCH_NOSRC) {
+                continue;
+            }
+            if (updated->arch != ARCH_NOARCH && solvable->arch != ARCH_NOARCH) {
+                continue;
+            }
```

Comment:
LGTM, this condition is really good one. Interesting thing is that this condition will still create incorrect result for packages with incompatible architecture.

When installed package is `pkgA-1-1.noarch` and available are `pkgA-2-1.x86_64` and `pkgA-2-1.s390` both will be considered as upgrades regardless of system architecture. Fortunately it requires 2 rare conditions. I do not think that we should resolve the issue, because for the final decision we have a solver.

Reply:
Well, although the `.s390` version is not installable, it still is (unlike the source rpm) a possible upgrade to a `.noarch` package. It depends on what we want from `filter_upgrades()` to exactly do... Btw we need to document this filter (I've just checked the `package_query.hpp` header file and found that the documentation is virtually missing).